### PR TITLE
test integration Google API

### PR DESCRIPTION
**Pull request by vldmrklnk**
GoogleSheets Api was added. GoogleController is a test controller.

### CODE REVIEW

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		public string ReadEntries()
+		{
+			GoogleCredential credential;
+
+			//var conf = _configuration.GetSection("GoogleCredentials").GetChildren().ToString();
+
+			//credential = GoogleCredential.FromJson(conf)
+			//		.CreateScoped(Scopes);
+
+			using (var stream = new FileStream("C:\\Users\\Vladimir\\Exadel\\InternshipBEOct2021\\InternshipBE\\WebApi\\client_secrets.json", FileMode.Open, FileAccess.Read))
```
Comment (Amarozka123) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: Move C:\\Users\\Vladimir\\Exadel\\InternshipBEOct2021\\InternshipBE\\WebApi\\client_secrets.json to config file

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		public GoogleSheetService(IConfiguration configuration)
+		{
+			_configuration = configuration;
+		}
+
+		public string ReadEntries()
+		{
+			GoogleCredential credential;
+
+			//var conf = _configuration.GetSection("GoogleCredentials").GetChildren().ToString();
```
Comment (Amarozka123) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: Remove commented code

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+					.CreateScoped(Scopes);
+			}
+
+			service = new SheetsService(new BaseClientService.Initializer()
+			{
+				HttpClientInitializer = credential,
+				ApplicationName = ApplicationName,
+			});
+
+			var range = $"{sheet}!A1:F15";
```
Comment (Amarozka123) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: This should be moved to DB or config. We should change this without rebuild app

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+using System.Threading.Tasks;
+
+namespace BL.Services
+{
+	public class GoogleSheetService : IGoogleSheetService
+	{
+		private readonly string[] Scopes = { SheetsService.Scope.Spreadsheets };
+		private readonly string ApplicationName = "ExampleCSharp";
+		private readonly string SpreadsheetId = "1PokdyP8f8tWaJlE0NQvF2taCKthzc-iy8lbWLubrAek";
+		private readonly string sheet = "Answers";
```
Comment (Amarozka123) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: It should be moved to config file or DB

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        private readonly IGoogleSheetConfig _sheetConfig;
+        private SheetsService service;
+
+		public GoogleSheetService(IGoogleSheetConfig sheetConfig)
+		{
+			_sheetConfig = sheetConfig;
+		}
+
+		public string ReadEntries()
+        {
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: bracket

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        private SheetsService service;
+
+		public GoogleSheetService(IGoogleSheetConfig sheetConfig)
+		{
+			_sheetConfig = sheetConfig;
+		}
+
+		public string ReadEntries()
+        { 
+            var credential = GoogleCredential.FromFile("client_secrets.json").CreateScoped(Scope);
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: hardcode, separate to const

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+            service = new SheetsService(new BaseClientService.Initializer()
+            {
+                HttpClientInitializer = credential,
+                ApplicationName = _sheetConfig.ApplicationName
+            });
+
+            var range = $"{_sheetConfig.Sheet}!A1:ZZ";
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: hardcode, separate to const

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+				if (values != null && values.Count > 0)
+				{
+				var options = new JsonSerializerOptions
+				{
+					Encoder = JavaScriptEncoder.Create(UnicodeRanges.BasicLatin, UnicodeRanges.Cyrillic),
+					WriteIndented = true
+				};
+				string json = JsonSerializer.Serialize(values, options);
+
+				return json;
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: everything went

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            service = new SheetsService(new BaseClientService.Initializer()
+            {
+                HttpClientInitializer = credential,
+                ApplicationName = _sheetConfig.ApplicationName
+            });
+
+            var range = $"{_sheetConfig.Sheet}!A1:ZZ";
+            var request = service.Spreadsheets.Values.Get(_sheetConfig.SpreadsheetId, range);
+
+            try
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: remove try catch

Comment (romanstat) [reply InternshipBE/BL/Services/GoogleSheetService.cs]: we have exceptionmiddleware

File: InternshipBE/WebApi/Controllers/GoogleController.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	public class GoogleController : ControllerBase
+	{
+		private readonly IGoogleSheetService _googleSheetService;
+		public GoogleController(IGoogleSheetService googleSheetService)
+		{
+			_googleSheetService = googleSheetService;
+		}
+
+		[HttpGet]
+		[AllowAnonymous]
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/GoogleController.cs]: for what AllowAnonymous

File: InternshipBE/WebApi/Controllers/GoogleController.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+		public IActionResult GetData()
+		{
+			var json = _googleSheetService.ReadEntries();
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/GoogleController.cs]: return Ok (_googleSheetService.ReadEntries());

Comment (romanstat) [reply InternshipBE/WebApi/Controllers/GoogleController.cs]: async await?

**Comment by romanstat** [comment]
Need to save to db as candidate

File: InternshipBE/DAL/Migrations/20211025190247_Add all entities.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace DAL.Migrations
+{
+    public partial class Addallentities : Migration
```
Comment (romanstat) [inline InternshipBE/DAL/Migrations/20211025190247_Add all entities.cs]: Empty migration?

Comment (vldmrklnk) [reply InternshipBE/DAL/Migrations/20211025190247_Add all entities.cs]: It's not our migration. We pulled code from develop branch to our.

File: InternshipBE/BL/Mapping/Profiles/CandidateProfile.cs [hunk 13]
```diff
@@ -1,14 +1,57 @@
...
+				.ForPath(x => x.RegistationDate, o => o.MapFrom(src => src.RegistationDate))
+				.ForPath(x => x.IsPlanningToJoin, o => o.MapFrom(src => src.IsPlanningToJoin))
+				.ForPath(x => x.Links, o => o.MapFrom(src => src.Links))
+				.ForPath(x => x.Phone, o => o.MapFrom(src => src.Phone))
+				.ForPath(x => x.PrimarySkill, o => o.MapFrom(src => src.PrimarySkill))
+				.ForPath(x => x.ProfessionalCertificates, o => o.MapFrom(src => src.ProfessionalCertificates))
+				.ForPath(x => x.Skype, o => o.MapFrom(src => src.Skype))
+				.ForPath(x => x.StackType, o => o.MapFrom(src => src.StackType))
+				.ForPath(x => x.OtherInfo, o => o.MapFrom(src => src.OtherInfo))
+				.ForAllOtherMembers(x => x.Ignore());
```
Comment (romanstat) [inline InternshipBE/BL/Mapping/Profiles/CandidateProfile.cs]: change

File: InternshipBE/DAL/Database/ApplicationDbContext.cs [hunk 14]
```diff
@@ -9,8 +9,6 @@ public class ApplicationDbContext : IdentityDbContext<User>
 		public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : base(options) 
 		{
 		}
-		public DbSet<User> Users { get; set; }
```
Comment (romanstat) [inline InternshipBE/DAL/Database/ApplicationDbContext.cs]: users removed?

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+				//When UnitOfWork will be added, we going to change this row.
+				//_candidateService.AddListAsync(candidates); 
+			}
+			else
+			{
+				throw new Exception("No data was found");
+			}
+		}
+
+		private IList<IList<object>> CheckNew()
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: rename

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+		private IMapper _mapper; 
+		private SheetsService service;
+
+		public GoogleSheetService(IGoogleSheetConfig sheetConfig, IMapper mapper)
+		{
+			_mapper = mapper;
+			_sheetConfig = sheetConfig;
+		}
+
+		public void SaveNewToDataBase()
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: rename

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+			{
+				HttpClientInitializer = credential,
+				ApplicationName = _sheetConfig.ApplicationName
+			});
+
+			var range = $"{_sheetConfig.Sheet}{_sheetConfig.RangeSettings}";
+			var request = service.Spreadsheets.Values.Get(_sheetConfig.SpreadsheetId, range);
+
+			var response = request.Execute();
+			var values = response.Values;
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs:73]: .Skip(1)?

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+			{
+				var models = _mapper.Map(values, new List<CandidateDTO>());
+				var candidates = _mapper.Map(models, new List<Candidate>());
+
+				//When UnitOfWork will be added, we going to change this row.
+				//_candidateService.AddListAsync(candidates); 
+			}
+			else
+			{
+				throw new Exception("No data was found");
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: no new users were found

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+			}
+		}
+
+		private IList<IList<object>> CheckNew()
+		{
+			var values = GetValuesFromTable();
+			values.RemoveAt(0);
+
+			//When UnitOfWork will be added, we going to change this row.
+			//var candidates = _genericCandidate.GetAll().Result;
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: optimize, example countasync

File: InternshipBE/BL/DTOs/CandidateDTO.cs [hunk 20]
```diff
@@ -1,8 +1,14 @@
-﻿namespace BL.DTOs
+﻿using Shared.Enums;
+using System;
+
+namespace BL.DTOs
 {
     public class CandidateDTO
     {
         public int Id { get; set; }
+        public DateTime RegistationDate { get; set; }
```
Comment (romanstat) [inline InternshipBE/BL/DTOs/CandidateDTO.cs:10]: space

File: InternshipBE/BL/DTOs/CandidateDTO.cs [hunk 21]
```diff
@@ -27,9 +35,10 @@ public class CandidateDTO
         public string CurrentJob { get; set; }
 
         public string ProfessionalCertificates { get; set; }
+        public bool IsPlanningToJoin { get; set; }
 
-        public string BestContactTime { get; set; }
 
         public int TestTaskEvaluation { get; set; }
+        public DateTime BestContactTime { get; set; }
```
Comment (romanstat) [inline InternshipBE/BL/DTOs/CandidateDTO.cs]: spaces

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+		public GoogleSheetService(IGoogleSheetConfig sheetConfig, IMapper mapper, ICandidateRepository candidateRepository)
+		{
+			_candidateRepository = candidateRepository;
+			_mapper = mapper;
+			_sheetConfig = sheetConfig;
+		}
+
+		public async Task SaveNewCandidates()
+		{
+			var values = CheckCandidates().Result;
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: await

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace BL.Services
+{
+	public class GoogleSheetService : IGoogleSheetService
+	{
+		private readonly string Scope = SheetsService.Scope.Spreadsheets;
+		private readonly IGoogleSheetConfig _sheetConfig;
+		private readonly ICandidateRepository _candidateRepository;
+		private IMapper _mapper;
+		private SheetsService service;
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: _sheetsService

File: InternshipBE/Shared/Config/Interfaces/IGoogleSheetConfig.cs [hunk 24]
```diff
@@ -0,0 +1,15 @@
+﻿namespace Shared.Config.Interfaces
+{
+	public interface IGoogleSheetConfig
+	{ 
+		public  string ApplicationName { get; }
```
Comment (romanstat) [inline InternshipBE/Shared/Config/Interfaces/IGoogleSheetConfig.cs]: aling

File: InternshipBE/DAL/Repositories/CandidateRepository.cs [hunk 25]
```diff
@@ -8,17 +8,23 @@
...
+		{
+			var internship = await _context.Internships.Include(x => x.Candidate).FirstOrDefaultAsync(x => x.Id == id);
 
-            return internship?.Candidate.Skip(itemsCount * pageNumber).Take(itemsCount).ToList();
-        }
-    }
+			return internship?.Candidate.Skip(itemsCount * pageNumber).Take(itemsCount).ToList();
+		}
+
+		public async Task SaveListCandidatesAsync(List<Candidate> candidates)
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/CandidateRepository.cs]: add generic method for range save

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+
+				await _candidateRepository.SaveListCandidatesAsync(candidates);
+			}
+			else
+			{
+				await Task.Run(() => throw new Exception("No new candidates was found"));
+			}
+		}
+
+		private async Task<IList<IList<object>>> CheckCandidates()
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: CheckCandidates() and return List? and check for what? rename

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+		}
+
+		public async Task SaveNewCandidates()
+		{
+			var values = CheckCandidates().Result;
+
+			if (values != null)
+			{
+				var models = _mapper.Map(values, new List<CandidateDTO>());
+				var candidates = _mapper.Map(models, new List<Candidate>());
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: replace to code convention

File: InternshipBE/BL/Services/GoogleSheetService.cs [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+			if (values != null)
+			{
+				var models = _mapper.Map<List<CandidateDTO>>(values);
+				var candidates = _mapper.Map<List<Candidate>>(models);
+
+				await _candidateRepository.RangeSaveAsync(candidates);
+			}
+		}
+
+		private async Task<IList<IList<object>>> CheckNewCandidatesAsync()
```
Comment (romanstat) [inline InternshipBE/BL/Services/GoogleSheetService.cs]: Maybe GetNew...?

**Review by romanstat — APPROVED** [review]

**Review by romanstat — APPROVED** [review]

**Review by Xodoc — APPROVED** [review]

**Review by Evgeniy-Sinicin — APPROVED** [review]

**Pull request merged**